File: 0623-add-one-row-to-tree/0623-add-one-row-to-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode AddOneRow(TreeNode root, int val, int depth) {
        if(depth == 1){
            return new TreeNode(val, root, null);
        }
        
        Helper(root, val, depth);
        
        return root;
    }
    
    private void Helper(TreeNode curr, int val, int depth){
        if(depth == 2){
            TreeNode newLeft = new TreeNode(val, curr.left, null);
            TreeNode newRight = new TreeNode(val, null, curr.right);
            curr.left = newLeft;
            curr.right = newRight;
        }else{
            if(curr.left != null){
                Helper(curr.left, val, depth - 1);
            }
            
            if(curr.right != null){
                Helper(curr.right, val, depth - 1);
            }
        }
    }
}

File: 0623-add-one-row-to-tree/0623-add-one-row-to-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode addOneRow(TreeNode root, int v, int d) {
        return dfs(root, v, d, 0);
    }
    
    private TreeNode dfs(TreeNode node, int v, int d, int dir) {
        if (d == 1) {
            TreeNode res = new TreeNode(v);
            if (dir == 0) res.left = node;
            else res.right = node;
            return res;
        }
        if (node == null) return node;
        node.right = dfs(node.right, v, d - 1, 1);
        node.left = dfs(node.left, v, d - 1, 0);
        return node;
    }
}